Runtime/Core/TargetPositionCache.cs

```diff
@@ -0,0 +1,244 @@
+using UnityEngine;
+using System.Collections.Generic;
+using System;
+
+namespace Cinemachine
+{
+    internal class TargetPositionCache
+    {
+        public static bool UseCache { get; set; }
+        public enum Mode { Disabled, Record, Playback }
+        public static Mode m_CacheMode = Mode.Disabled;
+
+        public static Mode CacheMode
+        {
+            get => m_CacheMode;
+            set
+            {
+                if (value == m_CacheMode)
+                    return;
+                m_CacheMode = value;
+                switch (m_CacheMode)
+                {
+                    default: case Mode.Disabled: ClearCache(); break;
+                    case Mode.Record: InitCache(); break;
+                    case Mode.Playback: CreatePlaybackCurves(); break;
+                }
+            }
+        }
+
+        public static float CurrentTime { get; set; }
+
+        class CacheEntry
+        {
+            public AnimationCurve X = new AnimationCurve();
+            public AnimationCurve Y = new AnimationCurve();
+            public AnimationCurve Z = new AnimationCurve();
+            public AnimationCurve RotX = new AnimationCurve();
+            public AnimationCurve RotY = new AnimationCurve();
+            public AnimationCurve RotZ = new AnimationCurve();
+
+            struct RecordingItem : IComparable<RecordingItem>
+            {
+                public float Time;
+                public Vector3 Pos;
+                public Vector3 Rot;
+                public int CompareTo(RecordingItem other) { return Time.CompareTo(other.Time); }
+            }
+            List<RecordingItem> RawItems = new List<RecordingItem>();
+
+            const float kResolution = 0.033f;
+
+            public void AddRawItem(float time, Transform target)
+            {
+                var n = RawItems.Count;
+                if (n == 0 || Mathf.Abs(RawItems[n-1].Time - time) >= kResolution)
+                    RawItems.Add(new RecordingItem
+                    {
+                        Time = time,
+                        Pos = target.position,
+                        Rot = target.rotation.eulerAngles
+                    });
+            }
+
+            public void CreateCurves()
+            {
+                var XList = new List<Keyframe>();
+                var YList = new List<Keyframe>();
+                var ZList = new List<Keyframe>();
+                var RotXList = new List<Keyframe>();
+                var RotYList = new List<Keyframe>();
+                var RotZList = new List<Keyframe>();
+
+                RawItems.Sort();
+                float time = float.MaxValue;
+                for (int i = 0; i < RawItems.Count; ++i)
+                {
+                    var item = RawItems[i];
```

Comment:
foreach (var item in RawItems)
{

Foreach may be nicer, but it's not necessarily better.

Reply:
I used to use foreach, but then discovered that there are situations where it does GC allocs.  So now I always do it this way

Reply:
Thanks for the info! :)